tests/unit/test_udf.py

```diff
@@ -0,0 +1,34 @@
+#
+# Copyright (c) 2012-2022 Snowflake Computing Inc. All rights reserved.
+#
+
+from unittest import mock
+
+import pytest
+
+from snowflake.connector import ProgrammingError
+from snowflake.snowpark import Session
+from snowflake.snowpark.exceptions import SnowparkSQLException
+from snowflake.snowpark.functions import udf
+from snowflake.snowpark.types import IntegerType
+from snowflake.snowpark.udf import UDFRegistration
+
+
+@mock.patch("snowflake.snowpark.udf.cleanup_failed_permanent_registration")
+def test_do_register_sp_negative(cleanup_registration_patch):
+    fake_session = mock.create_autospec(Session)
+    fake_session.get_fully_qualified_current_schema = mock.Mock(
+        return_value="database.schema"
+    )
+    fake_session._run_query = mock.Mock(side_effect=ProgrammingError())
+    fake_session.udf = UDFRegistration(fake_session)
+    with pytest.raises(SnowparkSQLException) as ex_info:
+        udf(lambda: 1, session=fake_session, return_type=IntegerType(), packages=[])
+    assert ex_info.value.error_code == "1304"
+    cleanup_registration_patch.assert_called()
+
+    fake_session._run_query = mock.Mock(side_effect=BaseException())
```

Comment:
Let's add some dummy error message to the exception and try to match in in `pytest.raises`? `BaseException` is a bit too broad for `pytest.raises`.